src/core/UbqScaledImageGenerator.cpp

```diff
@@ -0,0 +1,155 @@
+                                               supportedDimensions.height() > maxImageSize;
+        bool supportedDimsHaveSufficientSize =
+                supportedDimensions.width() > (float(maxImageSize) * 0.75) ||
+                supportedDimensions.height() > (float(maxImageSize) * 0.75);
+        if (!supportedDimsExceedMaxImageSize && supportedDimsHaveSufficientSize) {
+            // Use codec-supported dimensions for faster decoding
+            info = info.makeDimensions(supportedDimensions);
+        } else {
+            // Use desired dimensions directly
+            const auto scaledDimensions =
+                    SkISize::Make(static_cast<int32_t>(std::lround(desiredScale * width)),
+                                  static_cast<int32_t>(std::lround(desiredScale * height)));
+            info = info.makeDimensions(scaledDimensions);
+        }
+    }
+
+    return info;
+}
+
+UbqScaledImageGenerator::UbqScaledImageGenerator(std::unique_ptr<SkCodec> codec,
+                                                 sk_sp<SkData> data,
+                                                 int maxImageSize)
+        : INHERITED(adjust_info(codec.get(), maxImageSize))
+        , fCodec(std::move(codec))
+        , fData(std::move(data))
+        , fMaxImageSize(maxImageSize) {}
+
+sk_sp<SkData> UbqScaledImageGenerator::onRefEncodedData() { return fData; }
+
+auto UbqScaledImageGenerator::needsScaling() const -> bool {
+    return fCodec->dimensions() != getInfo().dimensions();
+}
+
+static inline auto isValidDecode(SkCodec::Result& result) -> bool {
+    switch (result) {
+        case SkCodec::kSuccess:
+        case SkCodec::kIncompleteInput:
+        case SkCodec::kErrorInInput:
+            return true;
+        default:
+            return false;
+    }
+}
+
+bool UbqScaledImageGenerator::getPixels(const SkImageInfo& info,
+                                        void* pixels,
+                                        size_t rowBytes,
+                                        const SkCodec::Options* options) {
+    SkPixmap dst(info, pixels, rowBytes);
+
+    std::function<bool(const SkPixmap&)> decode;
+    if (!needsScaling()) {  // Image is returned "as is", no fallback
+        decode = [this, options](const SkPixmap& pm) {
+            SkCodec::Result result = fCodec->getPixels(pm, options);
+            return isValidDecode(result);
+        };
+    } else {  // Query scaled version
+        decode = [this, options, info](const SkPixmap& pm) {
+            // Ask codec to return image at requested size
+            SkCodec::Result getPixelsResult = fCodec->getPixels(pm, options);
+            if (isValidDecode(getPixelsResult)) {
+                return true;
+            } else {
+                // Fall back to decode at closest dimension followed by forced downscale
+                float limit = fMaxImageSize;
+                float width = fCodec->getInfo().width();
```

Comment:
I know it's paranoid, but I always fear the divide by 0
```suggestion
                float width = std::max(1,  fCodec->getInfo().width());
```

Reply:
Paranoid isn't always a bad idea 😅